.github/workflows/Build.yml

```diff
@@ -24,14 +46,29 @@ jobs:
       uses: actions/cache@v2
       with:
         path: ~/.platformio
-        key: ${{ runner.os }}-${{ hashFiles('**/lockfiles') }}
+        key: ${{ runner.os }}-${{ hashFiles('**/platformio.ini') }}
+    - name: Cache PlatformIO build
+      uses: actions/cache@v2
+      with:
+        path: .pio
+        key: pio-${{ runner.os }}-${{ matrix.project }}
+        restore-keys: |
+          pio-${{ runner.os }}-
+          pio-
```

Comment:
Might need to revisit this to get new caches created correctly. Now Sony was changed and as other files change as well we might lose cache hits.

Reply:
Just for the record, I completely have no clue what your saying. :-P Well, I understand what your saying, but I don't know what it _technically_ means. i.e. I have no clue how any of your science caching magic works. I say you're a witch. Witch! Burn him! 🔥

Reply:
This part extracts the `.pio` directory from a tar archive before build so that cached builds can be used.
As times goes on more and more of the build cache will become out of date.
Since the key already matches perfectly new archives are not created.

Nothing will (probably) break, but builds will get slower again.
If something breaks just pull out that cache from the yml.

I do intend to keep an eye on this and create some kind of fix.

> magic is just technology we don't understand

I don't know either, I just modified the existing yml that you created. - now which is the witch? :P

Reply:
> I do intend to keep an eye on this and create some kind of fix.

So if I understand you correctly, we are referencing a point-in-time object cache, which will grow outdated with every change.

Can we automate creating the cache at all? Say, a weekly or daily etc tasks with what ever the `master` branch is at?

Reply:
Yes, the key is only a string.
We could use a key with the commit hash, but that would generate new caches on each commit.
But adding year and week number is probably a really good balance between not creating it every time causing many files in cache and not having it to outdated.

Right now we don't have much data, my thinking is to do the revisit when we see this causing an impact.
In worst case we should see build times of 12-15 minutes instead of the 2-3 minutes we have now. (but that is still better than the 30 minutes we had before)

Reply:
There is `key:` and `restore-keys`
if key matches it is used and no new cache is created.
On non key match, restore-keys is searched for alternative, and a new cache is created.

by adding more conditions to the main key, new caches will be created.
Max 10GB of caches in total will be kept.

https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action

Reply:
sounds good to me. I'll leave it with you. ;-)